### ISSUE TITLE
Report bug of 'CLIPredict(const CLIParam& param)' in cli_main.cc 

### ISSUE THREAD

**Issue by qf6101** [dmlc/xgboost#2781]
For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

## Environment info
Operating System: Ubuntu 15.10 \n \l

Compiler: gcc (Ubuntu 5.2.1-22ubuntu2) 5.2.1 20151010

Package used (python/R/jvm/C++): C++

`xgboost` version used: master ff9180cd73f21c21b8852b010dd664966bb00970

If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`) 

ff9180cd73f21c21b8852b010dd664966bb00970

2. Logs will be helpful (If logs are large, please upload as attachment).

If you are using jvm package, please 

1. add [jvm-packages] in the title to make it quickly be identified
2. the gcc version and distribution
gcc (Ubuntu 5.2.1-22ubuntu2) 5.2.1 20151010

If you are using python package, please provide

1. The python version and distribution
2. The command to install `xgboost` if you are not installing from source

If you are using R package, please provide

1. The R `sessionInfo()`
2. The command to install `xgboost` if you are not installing from source

## Steps to reproduce

1. Edit the directories of data files in xgboost/demo/binary_classification/mushroom.conf as follows.

```
diff --git a/demo/binary_classification/mushroom.conf b/demo/binary_classification/mushroom.conf
index 435c9bf..35c5dfb 100644
--- a/demo/binary_classification/mushroom.conf
+++ b/demo/binary_classification/mushroom.conf
@@ -20,10 +20,10 @@ num_round = 2
 # 0 means do not save any model except the final round model
 save_period = 0
 # The path of training data
-data = "agaricus.txt.train"
+data = "../data/agaricus.txt.train"
 # The path of validation data, used to monitor training process, here [test] sets name of the validation set
-eval[test] = "agaricus.txt.test"
+eval[test] = "../data/agaricus.txt.test"
 # evaluate on training data as well each round
 eval_train = 1
 # The path of test data
-test:data = "agaricus.txt.test"
+test:data = "../data/agaricus.txt.test"
```

2. In command line execute: ../../xgboost mushroom.conf max_depth=6. This worked well and produced the model file.

```
[10:37:27] 6513x127 matrix with 143286 entries loaded from agaricus.txt.train
[10:37:27] 1611x127 matrix with 35442 entries loaded from ../data/agaricus.txt.test
[10:37:27] src/cli_main.cc:198: Loading data: 0.0522792 sec
[10:37:27] boosting round 0, 0 sec elapsed
[10:37:27] src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 20 extra nodes, 0 pruned nodes, max_depth=5
[10:37:27] [0]	test-error:0.000000	train-error:0.000614
[10:37:27] boosting round 1, 0.175581 sec elapsed
[10:37:27] src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 18 extra nodes, 0 pruned nodes, max_depth=5
[10:37:27] [1]	test-error:0.000000	train-error:0.000000
[10:37:27] update end, 0.327175 sec in all
```

3. In command line execute: ../../xgboost mushroom.conf task=pred model_in=0002.model. After that, I got a segmentation error as follows.

```
[10:37:44] 1611x127 matrix with 35442 entries loaded from ../data/agaricus.txt.test
[10:37:44] start prediction...
Segmentation fault
```

## What have you tried?

I checked the source file [src/cli_main.cc](https://github.com/dmlc/xgboost/blob/master/src/cli_main.cc#L321-L322) and swapped the order of configuration and load. After re-compiling the xgboost I think I solved the problem.

```
[10:47:20] 1611x127 matrix with 35442 entries loaded from ../data/agaricus.txt.test
[10:47:20] start prediction...
[10:47:20] writing prediction to pred.txt
```

```
diff --git a/src/cli_main.cc b/src/cli_main.cc
index c0b29fb..5c301b6 100644
--- a/src/cli_main.cc
+++ b/src/cli_main.cc
@@ -318,8 +318,8 @@ void CLIPredict(const CLIParam& param) {
   std::unique_ptr<Learner> learner(Learner::Create({}));
   std::unique_ptr<dmlc::Stream> fi(
       dmlc::Stream::Create(param.model_in.c_str(), "r"));
-  learner->Configure(param.cfg);
   learner->Load(fi.get());
+  learner->Configure(param.cfg);
 
   if (param.silent == 0) {
     LOG(CONSOLE) << "start prediction...";
```

**Comment by vaibhav93** [comment]
I faced similar error while running runexp.sh for demo/binary_classification. @qf6101's solution fixed it for me

**Comment by khotilov** [comment]
Fixed in #2795. Thanks for investigating.
